### ISSUE TITLE
"Insert ROM" dialog usability issues when dragging and dropping files

### ISSUE THREAD

**Issue by theNestruo** [openMSX/openMSX#1554]
> - Windows version: 10.0.19045.3754
> - OpenMSX version: 19.1-467-g993c367c0 [Release]

When dragging and dropping a file in a partially covered OpenMSX window:
![image](https://github.com/openMSX/openMSX/assets/12213477/69011e5e-e5e7-466b-a5d2-ab1bf25a0c34)

Current behaviour:
- The "Insert ROM" dialog is show behind the active window.
- Clicking on the main OpenMSX window closes the dialog.
- Clicking on the OpenMSX window title bar brings the main OpenMSX window to the front (hiding the dialog, which is kept in the back).

Expected behaviour:
- The "Insert ROM" dialog is brought to the front.
- Clicking on the OpenMSX window does not hide or close the dialog.

**Comment by MBilderbeek** [comment]
I agree that it's a bit odd that the dialog can be closed by clicking on the openMSX window. Even if you have to make a choice (e.g. drive A or drive B), it still just closes and thus cancels the drop action.  Would you prefer that it can only be closed by closing it with a close button on the dialog?

When dropping in openMSX and having the dialog show, the dialog is part of the openMSX window (it's not separate). Would you expect the whole of openMSX to be put to front as well?

**Comment by theNestruo** [comment]
I think that dialog should be modal, just like the "Select ROM image for Cartridge Slot A" dialog is.

**Comment by MBilderbeek** [comment]
@m9710797 can we make the pop-up dialog modal? I think that works around the issue.
Currently in this situation you have to click the pop-up to both keep it shown and get the window raised.

**Comment by Vampier** [comment]
if rom is found in romdb skip the dialogue - all these extra pop up menus
are not needed when a rom is known

**Comment by m9710797** [comment]
@Vampier:
I partly agree. But only partly because you still may want to select the cartridge slot in which to insert the ROM.

@MBilderbeek, @theNestruo:
A modal dialog may indeed be a workaround. But two remarks:
* Then we do need to add a 'Cancel' button in all these popups (e.g. there is no such button when dropping a disk image).
* It does make this procedure more "visually invasive" (the whole MSX window changes color), and we did already get requests to make it less invasive (*1).

@MBilderbeek: could you experiment with this? Thanks.

--

(*1) I believe this request was in the context of streaming (on twitch.tv?) and you want the ability to change roms/disks without any visual feedback on the main MSX window?? Workaround for that use case could be to open the required "change-media" window(s) up-front, place them outside the main openMSX window, and when changing media click 'Apply' instead of 'OK' (=keep this window open).

**Comment by MBilderbeek** [comment]
@m9710797 about that last remark: yes, I know of at lesat one streamer (TheMP83) who would like to change contents of media outside the streamed openMSX main window. He finds drag and drop very comfortable to use, but it results in a 'greyed' openMSX main window (without anything else on it, as he dragged the actual pop up window off, and that location is saved, apparently). For him, it would already be helpful if the modal dialog would be visually less invasive, i.e. if (he could somehow configure that) it won't be visible in the main window.

**Comment by m9710797** [comment]
@MBilderbeek can you make a concrete suggestion for what to change?   (And feel free to experiment/change it yourself).

**Comment by MBilderbeek** [comment]
I am thinking about one of:
- make the greying of the main window for modal dialogs configurable, e.g. with a Boolean setting to enable/disable or to change the coloring
- don't use a modal dialog for dropping dialog

**Comment by Vampier** [comment]
Make it optional - I use it to onboard ROMs to the ROMDB and it speeds things up if I need to find the right mapper

**Comment by m9710797** [comment]
I don't know if it's possible to create a modal dialog without _greying_ the main window. And I also don't think it would be a good idea to do that ... for a modal dialog you do want some feedback that only that modal dialog will react to input now.

But feel free to experiment with a non-modal dialog for file-dropping.

And as suggested before:
* When the rom is present in the database, we don't need to ask for the rom-type.
* When there's only one possibility for the slot or only one disk-drive, we don't need to ask for that.

Maybe that can also eliminate the dialog in some cases.

**Comment by MBilderbeek** [comment]
@m9710797 I think we already take these cases into account, at least for the single disk drive. When only a single slot is present, you still get a pop up, where you get information and can override the mapper if you want.

<img width="544" height="237" alt="Image" src="https://github.com/user-attachments/assets/071c7e7e-4dd5-42b7-9162-fbeffb00703c" />

@theNestruo can you give an overview which of the usability issues you mentioned are still present? I have the feeling the thread has derailed a bit.

**Comment by MBilderbeek** [comment]
Oh and as far as I can see, the dialog isn't model in the current git HEAD.... I'm not sure if that was done intentionally, but it may change this issue.

**Comment by theNestruo** [comment]
> [@theNestruo](https://github.com/theNestruo) can you give an overview which of the usability issues you mentioned are still present? I have the feeling the thread has derailed a bit.

Still present in OpenMSX 21.0
